%\VignetteIndexEntry{Analyzing differential exon usage using RNA-seq}
%\VignettePackage{DESeq}

\documentclass{article}

\usepackage{Sweave}
\usepackage[a4paper]{geometry}
\usepackage{hyperref,graphicx}

 
\newcommand{\Robject}[1]{\texttt{#1}}
\newcommand{\Rpackage}[1]{\textit{#1}}
\newcommand{\Rclass}[1]{\textit{#1}}
\newcommand{\Rfunction}[1]{{\small\texttt{#1}}}

\title{\textsf{\textbf{Analyzing differential exon usage using RNA-seq}}}
\author{Simon Anders \\ Alejandro Reyes}

\begin{document}
\maketitle
\begin{abstract}
In this package we provide a method to systematically detect differential exon usage using RNA-seq. We use as input the number of reads falling on each of the exons of a genome. To give a demonstration of this Bioconductor package implementation we will use of a public data set from GEO Data set database under the record GSE18508 \footnote{Brooks AN, Yang L, Duff MO et Hansen KD et al. Conservation of an RNA regulatory map between Drosophila and mammals. \textit{Genome Research}, 2010}. For practicality and speed, this vignette only uses a subset of genes and only the paired-end sequenced lanes. Apart from the package providing the analysis of differential exon usage, we provide tools for the easy generation of the input and functions for the visualization and exploration of the test results.
\end{abstract}
\section{Starting with ExonSeq}
\subsection{Creating an ExonCountSet object}
An ExonCountSet object maintains the structure of eSet object with additions of other data fields (see code), the package provides accesses for each of the fields contained in an ExonCountSet object. The user can create an ExonCountSet object by providing directly a matrix with the counts of reads for each of the exons (as rows) of each sample (as columns) and two vectors of length equal to the row numbers of the matrix containing gene identifiers and exon identifiers.

\begin{Schunk}
\begin{Sinput}
> countfiles <- dir(system.file("files/", package = "ExonSeq"), 
+     pattern = "counts")
> lf <- lapply(countfiles, function(x) {
+     read.table(paste(system.file("files/", package = "ExonSeq"), 
+         x, sep = ""), header = FALSE)
+ })
> dcounts <- sapply(lf, function(x) {
+     x[, 2]
+ })
> rownames(dcounts) <- lf[[1]][, 1]
> dcounts <- dcounts[-which(rownames(dcounts) == "_empty" | rownames(dcounts) == 
+     "_ambiguous"), ]
> genesrle <- sapply(strsplit(as.character(rownames(dcounts)), 
+     ":"), "[[", 1)
> exonID <- paste("E", sapply(strsplit(rownames(dcounts), ":"), 
+     function(x) {
+         return(x[2])
+     }), sep = "")
> colnames(dcounts) <- sapply(strsplit(countfiles, ".counts", perl = TRUE), 
+     "[[", 1)
> head(dcounts)
\end{Sinput}
\begin{Soutput}
           treated2 treated3 untreated3 untreated4
SG10002:01        0        0          8          0
SG10002:02        0        3         11          0
SG10002:03       71       70         54         37
SG10002:04        0        0          0          0
SG10002:05     1434     1790       1358       1474
SG10002:06        0        0          0          0
\end{Soutput}
\begin{Sinput}
> head(genesrle)
\end{Sinput}
\begin{Soutput}
[1] "SG10002" "SG10002" "SG10002" "SG10002" "SG10002" "SG10002"
\end{Soutput}
\begin{Sinput}
> head(exonID)
\end{Sinput}
\begin{Soutput}
[1] "E01" "E02" "E03" "E04" "E05" "E06"
\end{Soutput}
\begin{Sinput}
> library(ExonSeq)
> ecs <- newExonCountSet(dcounts, c("treated", "treated", "untreated", 
+     "untreated"), geneIDs = genesrle, exonIDs = exonID)
> design(ecs)
\end{Sinput}
\begin{Soutput}
  treated2   treated3 untreated3 untreated4 
   treated    treated  untreated  untreated 
Levels: treated untreated
\end{Soutput}
\end{Schunk}
\normalsize
\subsection{ExonCountSet object via HTSeq}
Apart from the analysis for differential exon usage, ExonSeq provides a visualization function (\texttt{plotExonSeq}) and a HTML report generator (\texttt{ExonSeqHTML}) for the visualization and exploration of the results, but this requires information about the transcript annotation.  To this extend, the python HTSeq package provides the scripts aggregate\_genes.py and exon\_counts.py.  aggregate\_genes.py parses an annotation gtf file to define non-overlapping exonic regions: e.g. if a gene contains 2 exons that overlap, the script would define 2 exonic regions for the non-overlapping part of each exon and a third one for the overlapping part. It gives as output a second gtf file with the defined aggregated exonic regions.  The script exon\_counts.py takes the gtf file provided by aggregate\_genes.py and an alignment in sam format and counts the number of reads falling in each of the defined exonic regions. \\ The ExonSeq function \texttt{read.HTSeqCounts} is able to read the output of these scripts and returns an ExonCountSet object with the proper information to make the analysis for differential exon usage and generate the visualization of the results. If preferred, the user could also insert the annotation information manually to the ExonCountSet object directly after creating it (see functions \texttt{featureData, addTranscripts, showTranscripts}).

\begin{Schunk}
\begin{Sinput}
> countfiles <- dir(system.file("files/", package = "ExonSeq"), 
+     pattern = "counts")
> aggregatefile <- dir(system.file("files/", package = "ExonSeq"), 
+     pattern = "aggregate")
> aggregatefile <- paste(system.file("files/", package = "ExonSeq"), 
+     aggregatefile, sep = "")
> countfiles <- paste(system.file("files/", package = "ExonSeq"), 
+     countfiles, sep = "")
> countfiles
\end{Sinput}
\begin{Soutput}
[1] "/home/alejandro/software/opt/R-2.13/library/ExonSeq/files/treated2.counts"  
[2] "/home/alejandro/software/opt/R-2.13/library/ExonSeq/files/treated3.counts"  
[3] "/home/alejandro/software/opt/R-2.13/library/ExonSeq/files/untreated3.counts"
[4] "/home/alejandro/software/opt/R-2.13/library/ExonSeq/files/untreated4.counts"
\end{Soutput}
\begin{Sinput}
> aggregatefile
\end{Sinput}
\begin{Soutput}
[1] "/home/alejandro/software/opt/R-2.13/library/ExonSeq/files/aggregates_sample.gff"
\end{Soutput}
\begin{Sinput}
> ecs <- read.HTSeqCounts(countfiles, c("treated", "treated", "untreated", 
+     "untreated"), aggregatefile = aggregatefile)
> ecs
\end{Sinput}
\begin{Soutput}
ExonCountSet (storageMode: environment)
assayData: 8724 features, 4 samples 
  element names: counts 
protocolData: none
phenoData
  sampleNames:
    /home/alejandro/software/opt/R-2.13/library/ExonSeq/files/treated2
    /home/alejandro/software/opt/R-2.13/library/ExonSeq/files/treated3
    /home/alejandro/software/opt/R-2.13/library/ExonSeq/files/untreated3
    /home/alejandro/software/opt/R-2.13/library/ExonSeq/files/untreated4
  varLabels: sizeFactor condition
  varMetadata: labelDescription
featureData
  featureNames: SG10002:01 SG10002:02 ... SG9999:02 (8724 total)
  fvarLabels: geneID exonID ... strand (8 total)
  fvarMetadata: labelDescription
experimentData: use 'experimentData(object)'
Annotation:  
\end{Soutput}
\begin{Sinput}
> varMetadata(featureData(ecs))
\end{Sinput}
\begin{Soutput}
                               labelDescription
geneID     ID of gene to which the exon belongs
exonID      exon ID (unique only within a gene)
dispersion             exon dispersion estimate
pvalue                   pvalue from testForAIR
chr                          chromosome of exon
start                             start of exon
end                                 end of exon
strand                           strand of exon
\end{Soutput}
\begin{Sinput}
> head(featureData(ecs)$geneID)
\end{Sinput}
\begin{Soutput}
[1] SG10002 SG10002 SG10002 SG10002 SG10002 SG10002
1182 Levels: SG10002 SG10004 SG10007 SG1001 SG10020 SG1003 SG10056 ... SG9999
\end{Soutput}
\begin{Sinput}
> colnames(counts(ecs)) <- sampleNames(ecs) <- c("treated2", "treated3", 
+     "untreated3", "untreated4")
> head(counts(ecs))
\end{Sinput}
\begin{Soutput}
           treated2 treated3 untreated3 untreated4
SG10002:01        0        0          8          0
SG10002:02        0        3         11          0
SG10002:03       71       70         54         37
SG10002:04        0        0          0          0
SG10002:05     1434     1790       1358       1474
SG10002:06        0        0          0          0
\end{Soutput}
\end{Schunk}
\normalsize
\section{Size Factors and dispersion parameters}
Different samples might be sequenced with different depths, so we introduce size factors parameters in order to make the samples comparable.  ExonSeq uses the same implementation as in \texttt{DESeq} \footnote{Anders, S and Huber W. Differential expression analysis for sequence count data. Genome Biology, 2010} using the function \texttt{estimateSizeFactors}. The size factors are estimated as the median of the ratios of observed counts. 
\begin{Schunk}
\begin{Sinput}
> ecs <- estimateSizeFactors(ecs)
> sizeFactors(ecs)
\end{Sinput}
\begin{Soutput}
  treated2   treated3 untreated3 untreated4 
 0.9490972  1.1347877  0.9885038  0.9645947 
\end{Soutput}
\end{Schunk}
When testing for differential expression, one first need to have an estimate of the variance of the samples to distinguish between normal technical and biological variation (noise) from real effects of the different conditions in changes of gene expression.  However, biological replicates are necessary to estimate the dispersion between the samples, and this is not easy to estimate when the number of replicates is small. All the same applies for differential exon usage analysis.  In order to solve this problem, we make use of the Cox-Reid likelihood estimation of this parameters implemented by edgeR \footnote{Robinson Mark D. and McCarthy Davis J et al. edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics, 2009}, but with a slight modification at the time of testing. We estimate dispersions for each one of the exons as well as a common dispersion for all the exons, this is due to the fact that the common dispersion is not robust enough to manage with variance outliers, loosing control of type I error.  Because of this, we use the maximum between the common dispersion and the specific exon dispersion to test a specific exon. \\
The function \texttt{estimateDispersion} will call \texttt{estimateExonDispersionsForModelFrame} and \texttt{estimateCommonDispersion}, and it will store this parameters in their respectives fields in the ExonCountSet object.
\begin{Schunk}
\begin{Sinput}
> ecs <- estimateDispersions(ecs)
\end{Sinput}
\begin{Soutput}
Calculating model frames for each gene...
Estimating exon dispersions...
Estimating common dispersion...
disp =  0.00196	pll = -41667.517
disp =   0.0511	pll = -40955.077
disp =    0.383	pll = -46361.51
disp =   0.0122	pll = -39954.215
disp =   0.0124	pll = -39952.781
disp =   0.0132	pll = -39950.746
disp =   0.0132	pll = -39950.74
disp =   0.0132	pll = -39950.74
disp =   0.0132	pll = -39950.74
disp =   0.0132	pll = -39950.74
\end{Soutput}
\begin{Sinput}
> commonDispersion(ecs)
\end{Sinput}
\begin{Soutput}
[1] 0.01323773
\end{Soutput}
\begin{Sinput}
> head(featureData(ecs)$dispersion)
\end{Sinput}
\begin{Soutput}
[1] 9.040544628 7.904441252 0.009172978          NA 0.029289040          NA
\end{Soutput}
\end{Schunk}
\section{Testing for differential exon usage}
Having the dispersion estimates and the size factors, its possible to test for differential exon usage using a negative binomial distribution. For each gene we make a model frame with the function \texttt{modelFrameForGene} and use it to fit for every exon a generalized linear model with the interaction (formula=sample + exon + condition * I(exon == exonID)) and as a null model the glm without the interaction (formula=sample + exon + condition), then we compare the deviances of both regressions testing under a chi square distribution. All this is implemented in the function \texttt{testGeneForDEU}.  The function \texttt{testForDEU} will make a call to \texttt{testGeneForDEU} for all the genes, and will fill the featureData slots of the ExonCountSet object with the results.  The function \texttt{DEUresultTable} will give a summary of the results of the tests.
\begin{Schunk}
\begin{Sinput}
> head(modelFrameForGene(ecs, "SG7861"))
\end{Sinput}
\begin{Soutput}
    sample condition exon sizeFactor count
1 treated2   treated  E01  0.9490972   673
2 treated2   treated  E02  0.9490972    55
3 treated2   treated  E03  0.9490972   350
4 treated2   treated  E04  0.9490972   282
5 treated2   treated  E05  0.9490972   177
6 treated2   treated  E06  0.9490972   271
\end{Soutput}
\begin{Sinput}
> testGeneForDEU(ecs, "SG7861")
\end{Sinput}
\begin{Soutput}
        deviance df     pvalue
E01 2.003119e-01  1 0.65446921
E02 6.781141e-01  1 0.41023690
E03 1.318872e+00  1 0.25079451
E04 5.066361e+00  1 0.02439454
E05 2.051570e-01  1 0.65059004
E06 1.242826e-04  1 0.99110520
E07 1.088544e+00  1 0.29679395
E08 1.356150e-01  1 0.71267983
E09 5.286237e-03  1 0.94203967
E10 1.436671e+02  1 0.00000000
E11 2.530038e-01  1 0.61496791
E12 2.213986e+00  1 0.13676487
E13 7.277715e-01  1 0.39360649
E14 7.089488e-05  1 0.99328196
E15 1.167098e+00  1 0.27999836
E16 1.855847e-01  1 0.66661715
E17 9.515573e-01  1 0.32932320
E18 2.343210e+00  1 0.12583003
E19 1.051787e-01  1 0.74570133
E20 6.189633e+00  1 0.01285009
E21 8.814401e-01  1 0.34780756
E22 7.323036e-02  1 0.78669017
E23 4.950301e+00  1 0.02608613
\end{Soutput}
\begin{Sinput}
> ecs <- testForDEU(ecs)
\end{Sinput}
\begin{Soutput}
1000 genes out of 1182 processed
\end{Soutput}
\begin{Sinput}
> head(DEUresultTable(ecs))
\end{Sinput}
\begin{Soutput}
            geneID exonID dispersion     pvalue   padjust
SG10002:01 SG10002    E01 9.04054463 0.40810326 0.9081040
SG10002:02 SG10002    E02 7.90444125 0.74970879 0.9988187
SG10002:03 SG10002    E03 0.01323773 0.03613832 0.4116104
SG10002:04 SG10002    E04 0.01323773         NA        NA
SG10002:05 SG10002    E05 0.02928904 0.97518608 1.0000000
SG10002:06 SG10002    E06 0.01323773         NA        NA
\end{Soutput}
\end{Schunk}
\section{Visualization}
ExonSeq has a function to visualize the results of \texttt{testForDEU} with options for plotting the normalized counts for each of the exons or the coefficients estimates of the glm (Figure 1). It has also options for the visualization of the transcripts (Figure 2), which provides a good way to visualize the possible events of isoform regulation. \\

\begin{Schunk}
\begin{Sinput}
> plotExonSeq(ecs, "SG7861", cex.axis = 1.2, cex = 1.3, lwd = 1.5)
\end{Sinput}
\end{Schunk}
\begin{figure}
\centering
\includegraphics[width=\textwidth]{exonseq_vignette-plot1}
\caption{Plot indicating the coefficient estimate values from \texttt{testForDEU}, in red are indicated the exons that showed significance, meaning differential exon usage.}
\label{plot1}
\end{figure}
\normalsize
This is the option of visualizing the transcripts (Figure 2): \\

\begin{Schunk}
\begin{Sinput}
> plotExonSeq(ecs, "SG7861", displayTranscripts = TRUE, cex.axis = 1.2, 
+     cex = 1.3, lwd = 2, legend = TRUE)
\end{Sinput}
\end{Schunk}
\begin{figure}
\centering
\includegraphics[width=\textwidth]{exonseq_vignette-plot2}
\caption{The same as figure 1, but plotting it with the annotated transcripts.}
\label{plot2}
\end{figure}
\normalsize
Or visualize the counts normalized by the size factor (Figure 3): \\

\begin{Schunk}
\begin{Sinput}
> plotExonSeq(ecs, "SG7861", coefficients = FALSE, norCounts = TRUE, 
+     cex.axis = 1.2, cex = 1.3, lwd = 2)
\end{Sinput}
\end{Schunk}
\begin{figure}
\centering
\includegraphics[width=\textwidth]{exonseq_vignette-plot3}
\caption{Plot indicating the normalized count values in the exon for each of the samples.}
\label{plot3}
\end{figure}
The package also provides an HTML report generator implemented in the function \texttt{ExonSeqHTML}.  This report will generate a result table with links to plots for the significant results, allowing a more detailed exploration of the results. To see an example of it, visit the an example \link{http://www.embl.de/~reyes/Example/Example/testForDEU.html}. This report was generated using the code:
\begin{Schunk}
\begin{Sinput}
> ExonSeqHTML(ecs, path = "/home/alejandro/Work/Functions/ExonSeq/trunk/test/Prueba/", 
+     file = "testForDEU.html", sort = TRUE)
\end{Sinput}
\end{Schunk}
\section{Other useful functions}
You can create the example dataset for this vignette directly with the function \texttt{makeExampleExonCountSet()}, in case you want to reproduce this vignette. The user can also subset the ExonCountSet object by genes and create gene count tables.  This might be useful to give an input to other packages like DESeq.
\begin{Schunk}
\begin{Sinput}
> ecs2 <- makeExampleExonCountSet()
> ecs3 <- subsetByGenes(ecs2, sample(levels(geneIDs(ecs)), 100))
> ecs3
\end{Sinput}
\begin{Soutput}
ExonCountSet (storageMode: environment)
assayData: 703 features, 4 samples 
  element names: counts 
protocolData: none
phenoData
  sampleNames:
    /home/alejandro/software/opt/R-2.13/library/ExonSeq/files/treated2
    /home/alejandro/software/opt/R-2.13/library/ExonSeq/files/treated3
    /home/alejandro/software/opt/R-2.13/library/ExonSeq/files/untreated3
    /home/alejandro/software/opt/R-2.13/library/ExonSeq/files/untreated4
  varLabels: sizeFactor condition
  varMetadata: labelDescription
featureData
  featureNames: SG10007:01 SG10007:02 ... SG9875:08 (703 total)
  fvarLabels: geneID exonID ... strand (8 total)
  fvarMetadata: labelDescription
experimentData: use 'experimentData(object)'
Annotation:  
\end{Soutput}
\begin{Sinput}
> head(geneCountTable(ecs))
\end{Sinput}
\begin{Soutput}
        treated2 treated3 untreated3 untreated4
SG10002     1505     1863       1431       1511
SG10004      431      521        498        558
SG10007      174      235        254        236
SG1001       293      346        337        334
SG10020     2338     2671       2538       2255
SG1003         5        6          6          2
\end{Soutput}
\end{Schunk}
\end{document}
